src/main/java/com/amazon/aws/iot/greengrass/semver/Range.java

```diff
@@ -0,0 +1,486 @@
+            } else if (xp) {
+                ret = String.format(">=%s.%s.0%s <%s.%s.0-0", M, m, pr, M, Integer.parseInt(m) + 1);
+            }
+
+            return ret;
+        }
+
+        // ~, ~> --> * (any, kinda silly)
+        // ~2, ~2.x, ~2.x.x, ~>2, ~>2.x ~>2.x.x --> >=2.0.0 <3.0.0-0
+        // ~2.0, ~2.0.x, ~>2.0, ~>2.0.x --> >=2.0.0 <2.1.0-0
+        // ~1.2, ~1.2.x, ~>1.2, ~>1.2.x --> >=1.2.0 <1.3.0-0
+        // ~1.2.3, ~>1.2.3 --> >=1.2.3 <1.3.0-0
+        // ~1.2.0, ~>1.2.0 --> >=1.2.0 <1.3.0-0
+        static String replaceTildes(String comp) {
+            return Arrays.stream(S_PLUS.split(comp.trim(), -1)).map(FreeFunctions::replaceTilde)
+                    .collect(Collectors.joining(" "));
+        }
+
+        private static String replaceTilde(String s) {
+            Pattern r = SemVerRegexes.re.get(SemVerRegexes.t.get("TILDE"));
+            Matcher mm = r.matcher(s);
+            if (!mm.find()) {
+                return s;
+            }
+
+            String M = mm.group(1);
+            String m = mm.group(2);
+            String p = mm.group(3);
+            String pr = mm.group(4);
+
+            String ret;
+            if (isX(M)) {
+                ret = "";
+            } else if (isX(m)) {
+                ret = String.format(">=%s.0.0 <%s.0.0-0", M, Integer.parseInt(M) + 1);
+            } else if (isX(p)) {
+                // ~1.2 == >=1.2.0 <1.3.0-0
+                ret = String.format(">=%s.%s.0 <%s.%s.0-0", M, m, M, Integer.parseInt(m) + 1);
+            } else if (pr != null) {
+                ret = String.format(">=%s.%s.%s-%s <%s.%s.0-0", M, m, p, pr, M,
+                        Integer.parseInt(m) + 1);
+            } else {
+                // ~1.2.3 == >=1.2.3 <1.3.0-0
+                ret = String.format(">=%s.%s.%s <%s.%s.0-0", M, m, p, M, Integer.parseInt(m) + 1);
+            }
+
+            return ret;
+        }
+
+        // ^ --> * (any, kinda silly)
+        // ^2, ^2.x, ^2.x.x --> >=2.0.0 <3.0.0-0
+        // ^2.0, ^2.0.x --> >=2.0.0 <3.0.0-0
+        // ^1.2, ^1.2.x --> >=1.2.0 <2.0.0-0
+        // ^1.2.3 --> >=1.2.3 <2.0.0-0
+        // ^1.2.0 --> >=1.2.0 <2.0.0-0
+        static String replaceCarets(String comp) {
+            return Arrays.stream(S_PLUS.split(comp.trim(), -1)).map(FreeFunctions::replaceCaret)
+                    .collect(Collectors.joining(" "));
+        }
+
+        static String replaceCaret(String c) {
+            Pattern r = SemVerRegexes.re.get(SemVerRegexes.t.get("CARET"));
+            Matcher mm = r.matcher(c);
+            if (!mm.find()) {
+                return c;
+            }
+
+            String M = mm.group(1);
+            String m = mm.group(2);
+            String p = mm.group(3);
+            String pr = mm.group(4);
+
+            String ret;
+            if (isX(M)) {
+                ret = "";
+            } else if (isX(m)) {
+                ret = String.format(">=%s.0.0 <%s.0.0-0", M, Integer.parseInt(M) + 1);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This code uses &#39;%s&#39; to format int: &#39;+&#39; expression. This is a potential locale-sensitive handling issue. It might cause errors in the handling and processing of the statement at line: 369. Consider formatting this data with &#39;%d&#39; instead.